consent/service/consent.go

```diff
@@ -0,0 +1,130 @@
+package service
+
+import (
+	"context"
+
+	"github.com/tidepool-org/platform/consent"
+	"github.com/tidepool-org/platform/consent/store/mongo"
+	"github.com/tidepool-org/platform/errors"
+	"github.com/tidepool-org/platform/page"
+	"github.com/tidepool-org/platform/pointer"
+	structuredMongo "github.com/tidepool-org/platform/store/structured/mongo"
+
+	mongoDriver "go.mongodb.org/mongo-driver/mongo"
+)
+
+type ConsentService struct {
+	bddpSharer              BigDataDonationProjectSharer
+	consentMailer           *ConsentMailer
+	consentRepository       *mongo.ConsentRepository
+	consentRecordRepository *mongo.ConsentRecordRepository
+	dbClient                *mongoDriver.Client
+}
+
+func NewConsentService(consentMailer *ConsentMailer, bddpSharer BigDataDonationProjectSharer, consentRepository *mongo.ConsentRepository, consentRecordRepository *mongo.ConsentRecordRepository, dbClient *mongoDriver.Client) consent.Service {
+	return &ConsentService{
+		bddpSharer:              bddpSharer,
+		consentMailer:           consentMailer,
+		consentRepository:       consentRepository,
+		consentRecordRepository: consentRecordRepository,
+		dbClient:                dbClient,
+	}
+}
+
+func (c *ConsentService) ListConsents(ctx context.Context, filter *consent.Filter, pagination *page.Pagination) (*structuredMongo.ListResult[consent.Consent], error) {
+	return c.consentRepository.List(ctx, filter, pagination)
+}
+
+func (c *ConsentService) EnsureConsent(ctx context.Context, consent *consent.Consent) error {
+	return c.consentRepository.EnsureConsent(ctx, consent)
+}
+
+func (c *ConsentService) GetConsentRecord(ctx context.Context, userID string, id string) (*consent.Record, error) {
+	return c.consentRecordRepository.GetConsentRecord(ctx, userID, id)
+}
+
+func (c *ConsentService) CreateConsentRecord(ctx context.Context, userID string, create *consent.RecordCreate) (*consent.Record, error) {
+	pagination := page.NewPagination()
+	pagination.Size = 1
+
+	res, err := structuredMongo.WithTransaction(ctx, c.dbClient, func(sessCtx mongoDriver.SessionContext) (any, error) {
+		consents, err := c.ListConsents(sessCtx, &consent.Filter{
+			Type:    pointer.FromAny(create.Type),
+			Version: pointer.FromAny(create.Version),
+		}, pagination)
+		if err != nil {
+			return nil, err
+		}
+		if len(consents.Data) == 0 {
+			return nil, errors.New("invalid consent type and version combination")
+		}
+		cons := consents.Data[0]
+
+		records, err := c.consentRecordRepository.ListConsentRecords(sessCtx, userID, &consent.RecordFilter{
+			Latest: pointer.FromAny(true),
+			Status: pointer.FromAny(consent.RecordStatusActive),
+			Type:   pointer.FromAny(create.Type),
+		}, pagination)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(records.Data) > 0 {
+			if create.Type == consent.TypeBigDataDonationProject {
+				err = c.bddpSharer.Unshare(ctx, userID)
```

Comment:
Is this necessary? The method is named CreateConsentRecord, implying that the user is granting consent, so we shouldn't need to unshare then reshare, should we?

If we can remove this, then we prevent a case were a user goes to grant consent to a new version, and in doing so this method unshares, but then a database error causes an abort on line 88, and then the transaction fails, but the user is still unshared. We have no record of the revocation attempt, but the user is mysteriously unshared, even when their INTENT was to keep sharing.

Reply:
You're absolutely right. I reorganized the code to avoid the edge case you described.